message_header: Add date on every recipient bar in search results.

abdelrahman725 opened the pull request:
Fixes #31958

[CZO Discussion ](https://chat.zulip.org/#narrow/stream/101-design/topic/Date.20indicators.20in.20search.20results)

<details>
  <summary>See Before</summary>

  ![before_adding_date](https://github.com/user-attachments/assets/90fe0ffe-ff19-4cbe-bc39-5bca4ba5acfa)
</details>

<details>
  <summary>See After</summary>

![after_adding_date](https://github.com/user-attachments/assets/3102b7e4-f315-4924-b21e-d6187cdeeee1)
</details>

abdelrahman725 commented on web/src/message_list_view.ts:
```diff
@@ -159,6 +161,16 @@ function same_recipient(a: MessageContainer | undefined, b: MessageContainer | u
 }
 
+function is_search_view(): boolean {
+    const path = window.location.hash.split("/");
+    // The following check ensures the user is actually searching for something,
+    // and not to confuse it with the case in which the channel or topic name is "search".
+    if (path.length >= 2 && path.at(-2) === "search") {
+        return true;
+    }
```
### This matches the following cases : 

- `/#narrow/search/foo`
- `/#narrow/stream/stream_name/search/foo`
- `/#narrow/stream/search/topic/search/search/foo` 
- `/#narrow/stream/search/topic/search/search/search` both channel and topic are named "search", and the user is searching the string "search"

### But doesn't match the following : 

- `/#narrow/stream/stream_name/topic/search`
- `/#narrow/stream/search/topic/search`

timabbott replied:
Use a function in `filter.ts` to check this, rather than ad-hoc parsing the URL in another place.

abdelrahman725 replied:
I see `filter.ts` responsible for the searching functionality but nothing there checks if current view (e.g. current URL) is a search view

abdelrahman725 replied:
Instead I will move my function to `hash_util.ts`

timabbott commented on web/src/message_list_view.ts:
```diff
@@ -114,6 +115,7 @@ export type MessageGroup = {
           is_private: true;
           pm_with_url: string;
           recipient_users: RecipientRowUser[];
+          is_searched_messages: boolean;
```
`is_search_view` would be a more natural name for this. Even better would be to negate the meaning and just say what it means, and do a `always_display_date` variable.

abdelrahman725 commented:
Addressed the Review comments.

Since I didn't yet find an existing utility which checks if the current view a search results view, I just  moved `is_search_view()` to `hash_util.ts`.

alya commented:
@amanagr Could you please review this one? I haven't tested.

timabbott commented on web/src/hash_util.ts:
```diff
@@ -325,3 +325,13 @@ export function decode_stream_topic_from_url(
+        return true;
+    }
+    return false;
+}
```
No, you cannot parse the URL again.

Think a bit more. There's already logic, probably in `message_list_view.ts`, that determines whether to show individual cards for each message. How does that determine whether it's a search view? Your new logic should use the same method.

abdelrahman725 replied:
ok

timabbott commented on web/src/message_list_view.ts:
```diff
@@ -425,6 +427,14 @@ function maybe_restore_focus_to_message_edit_form(): void {
+        return true;
+    }
+    return false;
+}
```
I think you want to use `supports_collapsing_recipients` instead, since views like "starred messages" are also search views.

abdelrahman725 commented:
## Updates :  
use `supports_collapsing_recipients()` instead

now feature applies to the following view :

1. Search results
2. Senty by `user_name`
3. Starred messages
4. Reactions

timabbott commented:
OK, this version looks great, merged, thanks @abdelrahman725!

Hopefully this has been a useful lesson in trying to figure out the precisely right logic; our goal is for it to not require like 4 round trips for a fairly small change like this one.
